File: app/src/main/java/eu/darken/bluemusic/bluetooth/core/SourceDeviceWrapper.java
package eu.darken.bluemusic.bluetooth.core;

import android.bluetooth.BluetoothClass;
import android.bluetooth.BluetoothDevice;
import android.os.Parcel;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.lang.reflect.Method;
import java.util.Locale;

import eu.darken.bluemusic.main.core.audio.AudioStream;
import timber.log.Timber;

class SourceDeviceWrapper implements SourceDevice {
    private final BluetoothDevice realDevice;

    SourceDeviceWrapper(BluetoothDevice realDevice) {
        this.realDevice = realDevice;
    }

    @Override
    public String getLabel() {
        String label = getAlias();
        if (label == null) label = getName();
        if (label == null) label = getAddress();
        return label;
    }

    @Override
    public boolean setAlias(String alias) {
        try {
            //noinspection JavaReflectionMemberAccess
            Method method = realDevice.getClass().getMethod("setAlias", String.class);
            return (boolean) method.invoke(realDevice, alias);
        } catch (Exception e) {
            Timber.e(e);
        }
        return false;
    }

    @Nullable
    @Override
    public String getAlias() {
        try {
            //noinspection JavaReflectionMemberAccess
            Method method = realDevice.getClass().getMethod("getAliasName");
            return (String) method.invoke(realDevice);
        } catch (Exception e) {
            Timber.e(e);
        }
        return null;
    }

    @Nullable
    @Override
    public String getName() {
        return realDevice.getName();
    }

    @NonNull
    @Override
    public String getAddress() {
        return realDevice.getAddress();
    }

    @Nullable
    @Override
    public BluetoothClass getBluetoothClass() {
        return realDevice.getBluetoothClass();
    }

    @Override
    public AudioStream.Id getStreamId(AudioStream.Type type) {
        switch (type) {
            case MUSIC:
                return AudioStream.Id.STREAM_MUSIC;
            case CALL:
                return AudioStream.Id.STREAM_BLUETOOTH_HANDSFREE;
            case RINGTONE:
                return AudioStream.Id.STREAM_RINGTONE;
            case NOTIFICATION:
                return AudioStream.Id.STREAM_NOTIFICATION;
            default:
                throw new IllegalArgumentException("Unsupported AudioStreamType: " + type);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        SourceDeviceWrapper that = (SourceDeviceWrapper) o;

        return getAddress().equals(that.getAddress());
    }

    @Override
    public int hashCode() {
        return getAddress().hashCode();
    }

    @Override
    public String toString() {
        return String.format(Locale.US, "Device(name=%s, address=%s)", getName(), getAddress());
    }


    protected SourceDeviceWrapper(Parcel in) {
        realDevice = in.readParcelable(BluetoothDevice.class.getClassLoader());
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(realDevice, flags);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<SourceDeviceWrapper> CREATOR = new Creator<SourceDeviceWrapper>() {
        @Override
        public SourceDeviceWrapper createFromParcel(Parcel in) {
            return new SourceDeviceWrapper(in);
        }

        @Override
        public SourceDeviceWrapper[] newArray(int size) {
            return new SourceDeviceWrapper[size];
        }
    };
}

File: app/src/main/java/eu/darken/bluemusic/bluetooth/core/SourceDeviceWrapper.kt
package eu.darken.bluemusic.bluetooth.core

import android.Manifest
import android.annotation.SuppressLint
import android.bluetooth.BluetoothDevice
import androidx.annotation.RequiresPermission
import eu.darken.bluemusic.common.debug.logging.Logging.Priority.ERROR
import eu.darken.bluemusic.common.debug.logging.log
import eu.darken.bluemusic.common.debug.logging.logTag
import eu.darken.bluemusic.common.hasApiLevel
import eu.darken.bluemusic.devices.core.DeviceAddr
import eu.darken.bluemusic.monitor.core.audio.AudioStream
import kotlinx.parcelize.Parcelize

@Parcelize
data class SourceDeviceWrapper(
    override val address: DeviceAddr,
    val alias: String?,
    val name: String?,
    override val deviceType: SourceDevice.Type,
    override val isConnected: Boolean,
) : SourceDevice {

    override val label: String
        get() = alias ?: name ?: address

    override fun getStreamId(type: AudioStream.Type): AudioStream.Id = when (type) {
        AudioStream.Type.MUSIC -> AudioStream.Id.STREAM_MUSIC
        AudioStream.Type.CALL -> AudioStream.Id.STREAM_BLUETOOTH_HANDSFREE
        AudioStream.Type.RINGTONE -> AudioStream.Id.STREAM_RINGTONE
        AudioStream.Type.NOTIFICATION -> AudioStream.Id.STREAM_NOTIFICATION
        AudioStream.Type.ALARM -> AudioStream.Id.STREAM_ALARM
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as SourceDeviceWrapper
        return address == that.address
    }

    override fun hashCode(): Int = address.hashCode()

    override fun toString(): String = "Device(name=$name, address=$address, isConnected=$isConnected)"

    companion object {
        private val TAG = logTag("Bluetooth", "SourceDevice")

        @RequiresPermission(anyOf = [Manifest.permission.BLUETOOTH_CONNECT, Manifest.permission.BLUETOOTH])
        fun from(realDevice: BluetoothDevice, isConnected: Boolean) = SourceDeviceWrapper(
            address = realDevice.address,
            alias = run {
                if (hasApiLevel(30)) {
                    @SuppressLint("NewApi")
                    realDevice.alias
                } else {
                    try {
                        val method = realDevice.javaClass.getMethod("getAliasName")
                        method.invoke(realDevice) as String?
                    } catch (e: Exception) {
                        log(TAG, ERROR) { "Failed to use getAliasName(): ${e.message}" }
                        null
                    }
                }
            },
            name = realDevice.name,
            deviceType = realDevice.bluetoothClass.toType(),
            isConnected = isConnected
        )
    }
}